doc/template_definition.md

```diff
@@ -4,20 +4,63 @@ tags: [template]
 keywords: template, definition, code, java
 ---
 
-Spoon provides developers a way of writing code transformations called
-**code templates**. Those templates are statically type-checked, in
-order to ensure statically that the generated code will be correct.
+Spoon provides developers a way to define so called
+**Spoon templates**. **Spoon template** is a part of code (for example  
+expression, statement, block, method, constuctor, type member, 
+interface, type, ... any Spoon model subtree),
+where parts of that code may be **template parameters**.
 
-A Spoon template is a regular Java class that taken as input by the Spoon templating engine to perform a transformation.
+**Spoon template** can be used in two ways:
+
+A) **to generate new code**. The generated code is a copy of code
+of **Spoon template**, where each **template parameter** is substituted
+by it's value. We call this operation **Generating**.
+```java
+Factory spoonFactory = ...
+//build a Spoon template
+Pattern spoonTemplate = ... build a spoon template. For example for an method ...
+//define values for parameters
+Map<String, Object> parameters = new HashMap<>();
+parameters.put("methodName", "i_am_an_generated_method");
+//generate a code using spoon template and parameters
+CtMethod<?> generatedMethod = spoonTemplate.substituteSingle(spoonFactory, CtMethod.class, parameters);
+```
 This is summarized in Figure below. A Spoon template can be seen as a
 higher-order program, which takes program elements as arguments, and returns a
 transformed program. Like any function, a template can be used in different
 contexts and give different results, depending on its parameters.
 
 ![Overview of Spoon's Templating System]({{ "/images/template-overview.svg" | prepend: site.baseurl }})
 
-Definition of templates
------------------------
+B) **to search for a code**. The found code is same like code of **Spoon template**,
+where code on position of **template parameter** may be arbitrary and is copied
+as value of **template parameter**. We call this operation **Matching**.
+```java
```

Comment:
My previous comment is particulary true here: I did not really understood what you meant by: 

> where code on position of **template parameter** may be arbitrary and is copied
as value of **template parameter**.

it might be easier with an example code/template -> obtained matching

Reply:
I speak about template matching, that matched code may have any AST subtree on the position of template parameter - because template parameter is a kind of "variable"

Reply:
Sorry but I still don't understand what you exactly mean here: what do you exactly intend by "on the position of the template parameter"? 

Let's keep an example: my template is the following: 

```java
public int mymethod(int _myvar_) {
   return _anotherVar_;
}
```

and I got this code:

```java
public int foo(int x) {
  return x*2;
}

public int bar(int y) { 
  return foo(y);
}
```

Then the template match against the 2 methods, and the param `anotherVar` will point for both CtExpression `x*2` and `foo(y)` right? or as `y` is also pointed by `myvar` it won't be that simple?

Reply:
It matches twice. 
First match points to method "foo" and have parameters: 
```
_myvar_ = int x
_anotherVar_ = x*2
```
Second match points to method "bar" and have parameters: 
```
_myvar_ = int y
_anotherVar_ = foo(y)
```

Reply:
ok that's what I expected then.